### ISSUE TITLE
Typeguard and MyPy disagree about ContextManagers

### ISSUE THREAD

**Issue by prescod** [agronholm/typeguard#115]
Given this code:

```
from typing import ContextManager, Generator
from contextlib import contextmanager


@contextmanager
def foo() -> ContextManager[str]:
    yield "Hello"


@contextmanager
def bar() -> Generator[str, None, None]:
    yield "Hello"


def main():
    foo()
    bar()
```

Mypy dislikes `foo`

```
code_to_check.py:5: error: The return type of a generator function should be "Generator" or one of its supertypes
code_to_check.py:5: error: Argument 1 to "contextmanager" has incompatible type "Callable[[], ContextManager[str]]"; expected "Callable[..., Iterator[<nothing>]]"
```

Typeguard dislikes `bar`

```
  File "type_check.py", line 6, in <module>
    code_to_check.main()
  File "/Users/pprescod/code/CumulusCI/code_to_check.py", line 17, in main
    bar()
  File "/Users/pprescod/.venvs/CCI-37/lib/python3.7/site-packages/typeguard/__init__.py", line 811, in wrapper
    check_return_type(retval, memo)
  File "/Users/pprescod/.venvs/CCI-37/lib/python3.7/site-packages/typeguard/__init__.py", line 635, in check_return_type
    raise exc from None
  File "/Users/pprescod/.venvs/CCI-37/lib/python3.7/site-packages/typeguard/__init__.py", line 633, in check_return_type
    check_type('the return value', retval, memo.type_hints['return'], memo)
  File "/Users/pprescod/.venvs/CCI-37/lib/python3.7/site-packages/typeguard/__init__.py", line 568, in check_type
    check_type(argname, value, origin_type, memo)
  File "/Users/pprescod/.venvs/CCI-37/lib/python3.7/site-packages/typeguard/__init__.py", line 598, in check_type
    format(argname, qualified_name(expected_type), qualified_name(value)))
TypeError: type of the return value must be collections.abc.Generator; got contextlib._GeneratorContextManager instead
```

**Comment by agronholm** [comment]
Do you feel that typeguard is in the wrong about this then? Typeguard only looks at the type of the actual return value.

**Comment by agronholm** [comment]
Mypy does not seem to consider the decorator at all, so in my opinion, Mypy is in the wrong here. Comment here if you disagree, but for the moment I'm going to close this.

**Comment by cjolowicz** [comment]
Just hit this as well.

I must say that I think mypy is doing the right thing here. After all, a function decorator is just syntactic sugar for a function that takes another function as an argument. For example, take this code: 

```
from typing import Generator
from contextlib import contextmanager

def foo_generator() -> Generator[str, None, None]:
    yield "Hello"

foo = contextmanager(foo_generator)
```

Declare `foo_generator` to return `typing.ContextManager`, and `foo_generator` won't type-check in mypy or typeguard.

It will also have the wrong signature as an argument to `contextlib.contextmanager`, whose type signature is [this](https://github.com/python/typeshed/blob/5e76f51930190efa38d557253894e6584bce189d/stdlib/2and3/contextlib.pyi#L37):

```python
def contextmanager(func: Callable[..., Iterator[_T]]) -> Callable[..., _GeneratorContextManager[_T]]: ...
```

Unfortunately, the [implementation](https://github.com/python/cpython/blob/d4a63031dac0078d97677da43a1e3004ea64c993/Lib/contextlib.py#L211-L241) of `contextmanager` simply copies `__annotations__` to the decorated function, using `functools.wraps`. So this situation is not entirely typeguard's fault either.

**Comment by cjolowicz** [comment]
So one workaround would be the following:

```python
from typing import ContextManager, Generator
from contextlib import contextmanager

@contextmanager
def foo() -> Generator[str]:
    yield "Hello"

foo.__annotations__["return"] = ContextManager
```

I wonder if `contextlib.contextmanager` couldn't simply do this for us?

**Comment by cjolowicz** [comment]
This wrapper for `contextlib.contextmanager` is a more reusable workaround:

```python
import contextlib
from typing import Callable, ContextManager, Iterator, TypeVar

T = TypeVar("T")

def contextmanager(
    func: Callable[..., Iterator[T]]
) -> Callable[..., ContextManager[T]]:
    result = contextlib.contextmanager(func)
    result.__annotations__["return"] = ContextManager[T]
    return result
```

This should allow to use `@contextmanager` in a way that type-checks with both mypy and typeguard.

**Comment by cjolowicz** [comment]
There is an issue with the workaround posted above when both the decorated function and the original function are called in a program. `contextlib.contextmanager` (via `functools.wraps`) shallow-copies the type annotations from the original function. This means that the return type annotation was also applied to the original function. This will lead to a runtime type error when that function is called directly.

This issue can be fixed by updating the annotations non-destructively, like this:

```python
import contextlib
from typing import Callable, ContextManager, Iterator, TypeVar

T = TypeVar("T")

def contextmanager(
    func: Callable[..., Iterator[T]]
) -> Callable[..., ContextManager[T]]:
    result = contextlib.contextmanager(func)
    result.__annotations__ = {**func.__annotations__, "return": ContextManager[T]}
    return result
```

**Comment by agronholm** [comment]
Good news: this is now fixed in the 3.0 branch (493a732c).
